receiver/rabbitmqreceiver/internal/metadata/generated_metrics_v2.go

```diff
@@ -392,16 +396,73 @@ func NewMetricsBuilder(settings MetricsSettings, options ...metricBuilderOption)
+		mb.metricsCapacity = rm.InstrumentationLibraryMetrics().At(0).Metrics().Len()
+	}
+	if mb.resourceCapacity < rm.Resource().Attributes().Len() {
+		mb.resourceCapacity = rm.Resource().Attributes().Len()
+	}
+}
+
+// ResourceOption applies changes to provided resource.
+type ResourceOption func(pdata.Resource)
+
+// WithRabbitmqNodeNameResourceAttribute sets provided value as "rabbitmq.node.name" attribute for current resource.
+func WithRabbitmqNodeNameResourceAttribute(val string) ResourceOption {
+	return func(r pdata.Resource) {
+		r.Attributes().UpsertString("rabbitmq.node.name", val)
+	}
+}
+
+// WithRabbitmqQueueNameResourceAttribute sets provided value as "rabbitmq.queue.name" attribute for current resource.
+func WithRabbitmqQueueNameResourceAttribute(val string) ResourceOption {
+	return func(r pdata.Resource) {
+		r.Attributes().UpsertString("rabbitmq.queue.name", val)
+	}
+}
+
+// WithRabbitmqVhostNameResourceAttribute sets provided value as "rabbitmq.vhost.name" attribute for current resource.
+func WithRabbitmqVhostNameResourceAttribute(val string) ResourceOption {
+	return func(r pdata.Resource) {
+		r.Attributes().UpsertString("rabbitmq.vhost.name", val)
+	}
+}
+
+// EmitForResource saves all the generated metrics under a new resource and updates the internal state to be ready for
+// recording another set of data points as part of another resource. This function can be helpful when one scraper
+// needs to emit metrics from several resources. Otherwise calling this function is not required,
+// just `Emit` function can be called instead. Resource attributes should be provided as ResourceOption arguments.
+func (mb *MetricsBuilder) EmitForResource(ro ...ResourceOption) {
+	rm := pdata.NewResourceMetrics()
+	rm.Resource().Attributes().EnsureCapacity(mb.resourceCapacity)
+	for _, op := range ro {
+		op(rm.Resource())
+	}
+	ils := rm.InstrumentationLibraryMetrics().AppendEmpty()
+	ils.InstrumentationLibrary().SetName("otelcol/rabbitmqreceiver")
+	ils.Metrics().EnsureCapacity(mb.metricsCapacity)
+	mb.metricRabbitmqConsumerCount.emit(ils.Metrics())
+	mb.metricRabbitmqMessageAcknowledged.emit(ils.Metrics())
+	mb.metricRabbitmqMessageCurrent.emit(ils.Metrics())
+	mb.metricRabbitmqMessageDelivered.emit(ils.Metrics())
+	mb.metricRabbitmqMessageDropped.emit(ils.Metrics())
+	mb.metricRabbitmqMessagePublished.emit(ils.Metrics())
+	if ils.Metrics().Len() > 0 {
+		mb.updateCapacity(rm)
+		rm.MoveTo(mb.metricsBuffer.ResourceMetrics().AppendEmpty())
+	}
+}
+
+// Emit returns all the metrics accumulated by the metrics builder and updates the internal state to be ready for
+// recording another set of metrics. This function will be responsible for applying all the transformations required to
+// produce metric representation defined in metadata and user settings, e.g. delta or cumulative.
+func (mb *MetricsBuilder) Emit(options ...ResourceOption) pdata.Metrics {
+	mb.EmitForResource(options...)
```

Comment:
I really like this pattern overall but it isn't clear to me if it creates an extra resource when used like this:
```go
for _, resource := range resources {
    ...
    mb.EmitForResource(metadata.WithFooResourceAttribute(resource.foo))
}
mb.Emit()
```

I think this is a common pattern for collecting, but it looks like maybe it will create an empty resource at the end. Possibly I am missing how this is avoided but want to make sure.

Reply:
No empty resources created in this case. Metric builder always trims empty resources